Refactor with delegate

paikwiki opened the pull request:
Close #12

paikwiki committed:
Remove comment: "UI UPDATE"

paikwiki committed:
Create GameDelegate with test() method

paikwiki committed:
Refactor didTapNumber(_ sender:) and didTapReset(_ sender:)

- Apply delegate pattern for button actions
- Detach codes for UI update from didTapNumber(_ sender:) and didTapReset(_ sender:)

paikwiki commented:
`GameDelegate`에 `func showResult(gameResult: Bool)` 이렇게 작성했는데 객체끼리 주고 받는 게 객체가 아니라 값일 경우, 리팩토링의 대상일 수 있다고 들었었거든. 게임과 뷰컨트롤러가 부울 값을 주고 받는 부분, 이대로 괜찮을까요?! 아니면 예전처럼 GameResult를 Enum으로 만들거나, 다른 좋은 방법이 있을까요? 아님 이대로도 괜찮은 걸까요?

rinndash reviewed (changes requested):
`GameDelegate`는 Game에서 일어난 일을 다른 객체에(이 경우에는 ViewController)에게 알리기 위한 프로토콜이어야 돼

즉 `Game`은 `GameDelegate`에 게임과 관련된 이벤트가 일어났다는 사실만 알려주면 되고, 그 이벤트가 일어났을 때 어떤 동작을 할지는 `GameDelegate` 프로토콜을 conforms 하는 객체가 알아서 결정할 일이야.

그렇다면 지금 프로토콜에 선언된 함수 3개는 `Game`과 관련이 있다고 볼 수 있을까? 지금 세 함수 모두 `ViewController`라는 개념에 종속된 함수들이지 **숫자 야구**라는 게임과는 상관이 없지.

만약에 이닝이나 게임이 끝났을 때 UILabel을 업데이트하는게 아니라 데이타베이스에 게임 결과를 저장하는 클래스를 `Game`의 `delegate`로 설정했다면 그런 상황을 지금 함수로 표현할 수 있을까? 

```swift
class Game {
  weak var delegate: ViewController?
}
```

이렇게 뷰와 관련된 라벨이라던가 얼럿을 보여주는 개념이 Game이 포함된다면 굳이 `GameDelegate`라는 프로토콜을 만들 필요가 없이 저렇게 `ViewController`를 직접 참조로 가지고 있어도 충분히 구현할 수 있어.

굳이 델리게이트 패턴을 사용하는 이유는 저렇게 Delegate 패턴을 구현하는 객체에 어떠한 제한도 두지 않고 사용할 수 있게 하기 위해 사용하는 거라는 걸 잊지 마세요.

참고로 [`UITableView`](https://developer.apple.com/documentation/uikit/uitableview)와 [`UITableViewDelegate`](https://developer.apple.com/documentation/uikit/uitableviewdelegate) 문서 읽어 보면서 두 개가 어떤 관계인지 한 번 살펴보고, 델리게이트 프로토콜의 함수나 프로퍼티 이름을 짓는 컨벤션도 존재하니까 `UITableViewDelegate` 이름들도 한 번 잘 살펴보고 `GameDelegate` 프로토콜의 선언을 어떻게 수정하면 좋을 지 잘 생각해봐~

rinndash commented:
> GameDelegate에 func showResult(gameResult: Bool) 이렇게 작성했는데 객체끼리 주고 받는 게 객체가 아니라 값일 경우, 리팩토링의 대상일 수 있다고 들었었거든. 

이런 얘기는 어디서 들었는지는 모르겠지만 완전 틀린 얘기입니다. 델리게이트 프로토콜에서 쓰이는 함수가 참조 타입을 쓰건 값타입을 쓰건 아무 상관이 없습니다. 오히려 참조 관리를 할 필요 없는 값 타입을 쓰는 게 더 좋지요. 그리고 `enum` 역시 값 타입이니 `GameResult` 타입을 만들어 쓴다고 해서 달라지는 건 없습니다.

rinndash commented on NumberBaseball/ViewController.swift:
```diff
@@ -48,38 +49,34 @@ class ViewController: UIViewController {
-        // UI UPDATE
+    func updateUILabels() {
         inningViews[game.inningCount - 1].pitchesLabel.text = game.inning.description
         if game.inning.isEnded {
```
`GameDelegate`프로토콜을 잘 구현했다면 `ViewController`에서 `game.inning.isEnded`와 `game.inning.isOver`를 사용하지 않아도 됩니다.

paikwiki commented:
> > GameDelegate에 func showResult(gameResult: Bool) 이렇게 작성했는데 객체끼리 주고 받는 게 객체가 아니라 값일 경우, 리팩토링의 대상일 수 있다고 들었었거든.
> 
> 이런 얘기는 어디서 들었는지는 모르겠지만 완전 틀린 얘기입니다. 델리게이트 프로토콜에서 쓰이는 함수가 참조 타입을 쓰건 값타입을 쓰건 아무 상관이 없습니다. 오히려 참조 관리를 할 필요 없는 값 타입을 쓰는 게 더 좋지요. 그리고 `enum` 역시 값 타입이니 `GameResult` 타입을 만들어 쓴다고 해서 달라지는 건 없습니다.

저 이야기는 자바스크립트 강의에서 들었는데, "값" 대신 "값 객체"를 쓰는 게 좋다고 했던 것 같아.. 여튼 이 문제가 중요한 게 아니라 현재는 델리게이트가 목적대로 구현되지 않은 게 더 큰 문제로군...:sob:

paikwiki committed:
Change methods of GameDelegate

- updateUILabels() -> gameDidPitch(pitchString:) and gameShouldEndInning(inningResultString:)
- resetUILabels() -> gameShouldRestart()
- showResult(gameResult:) -> gameShouldEnd(gameResult:, answerString:)

paikwiki committed:
Change gameResult to isThreeStrikes(computed property)

- gameResult was a boolean type property of Game.
- 정답여부 확인을 위한 변수인데 Bool 타입의 gameResult라는 이름이 부적절하다고 판단했다.
- 게임이 끝났을 때 스트라이크 3개로 끝났는지 체크해주는 계산 프로퍼티로 변환
- 이에 따라 GameDelegate 메서드 파라미터 변경

paikwiki committed:
Change startNextInning() to finishCurrentInning() of Game class

- Game 클래스의 pitchABall(pitchNumber:)의 로직을 간결하게 하기 위해 변경했다.

paikwiki committed:
Rename a method: pitchABall(pitchNumber:) -> doPitching(pitchNumber)

paikwiki commented:
델리게이트에 대해서 짬짬이 공부하느라 커밋이 늦었네;; 일단, 직전에 만든 델리게이트 프로토콜이 왜 잘못됐는지는 확실히 깨달았는데, 지금 고친 게 괜찮은 건지는 확신이 없다. String을 넘기는 것도 뭔가 마음에 들지 않고... 지금으로선 이것이 최선이 아닌가 합니다.

rinndash commented on NumberBaseball/GameDelegate.swift:
```diff
@@ -0,0 +1,14 @@
+//  Copyright © 2020 Changhyun Baek. All rights reserved.
+//
+
+protocol GameDelegate: class {
```
정답은 없지만 나라면 이렇게 프로토콜을 짤 것 같아

```swift
protocol GameDelegate {
  func game(_ game: Game, didEndInning inning: Inning)
  func gameDidFinish(_ game: Game)
}
```
1. 내 생각엔 이 프로토콜에 `gameDidPitch`와 `gameShouldRestart`필요 없다고 생각해. `Game`의 `doPitching`과 `reset` 함수가 노출되어 있으니 그 함수를 호출하고 원하는 동작을 수행하면 되니 꼭 없어도 되는 함수라고 생각해. 반면, 이닝이 끝나는 것과 게임이 끝나는 것은 `Game` 바깥에서는 알 수 없고 `doPicthing`을 수행했을 때 해당 이벤트가 일어날지 아닐지 알 수 없으니 델리게이트를 이용해서 구현할 필요가 있지.

2. 그리고 델리게이트의 함수의 파라미터로 라벨에 그릴 문자열을 주고 있는데, 이건 모델에 속하는 `GameDelegate`가 라벨에 쓸 문자열에 대해 알고 있다는 이야기가 돼. 즉 모델과 뷰 간에 의존성이 존재한다는 말이지. GameDelegate는 모델에 속하는 정보를 넘겨주고, 이 정보를 어떻게 포맷팅할지는 뷰에 관한 일이니 ViewController, 나중에는 ViewModel이 결정하는 게 적절해.

rinndash replied:
그리고 델리게이트 프로토콜의 함수 이름도 짓는 컨벤션이 있으니 확인해 보세요. 
[UITableViewDelegate](https://developer.apple.com/documentation/uikit/uitableviewdelegate) 참고

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -7,43 +7,72 @@
     }
 
-    func pitchABall(pitchNumber: Int) {
+    func gameDidPitch(pitchString: String) {
```
특별한 이유가 없다면 델리게이트를 호출하는 함수를 이렇게 굳이 별도로 추출할 필요는 없어. 함수를 인라인 해주세요.

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -7,43 +7,72 @@
+        delegate?.gameDidPitch(pitchString: inning.description)
+    }
+
+    func gameShouldEndInning(inningResultString: String) {
```
함수 인라인

rinndash commented on NumberBaseball/Game.swift:
```diff
@@ -7,43 +7,72 @@
         }
+    }
 
+    func finishCurrentInning() {
```
`finishCurrentInning`이 함수는 `Game`에서만 쓰이고 있으니 `private`으로 설정해 주세요. 이 함수 외에도 Game 바깥에서 사용되지 않는 모든 프로퍼티와 함수는 `private` 으로 설정바랍니다.

paikwiki committed:
Add and remove methods of gameDelegate

- remove: gameShouldEndInning(inningResultString: String)
- remove: gameShoudRestart()
- remove: gameShouldEnd(isThreeStrkes: Bool, answerString: String)
- add: gameDidFinish(_ game: Game)
- add: game(_ game: Game, didEndInning inning: Inning, inningResultString: String)

paikwiki committed:
Update Podfile.lock

paikwiki commented:
`GameDelegate` 수정하는데 `game(_ game: Game, didEndInning inning: Inning, inningResultString: String)`로 작성해서, 권해줬던 메소드에다가 `inningResultString: String`라는 이상한 파라미터를 하나 추가했어;; 

Delegate를 적용하려고 하니까 화면에 출력하기 위한 문자열 처리가 말썽이네. `inningResultString`를 전달해주지 않고 처리하려면 `Game`의 위임객체 역할을 하는 `ViewController`가 `Answer`를 알고 있어야 하잖아. `ViewController`가 `judgePitching(inning: Inning)`을 부를 수 없으니 위임객체 역할을 하는 `ViewController`한테 문자열을 전달해주는 건데, 별로 좋지 않은 방법인데 뾰족한 방법을 잘 모르겠군요...

그나저나 이렇게 작은 앱인데도 메서드 몇 개 여기저기 고치기 시작하니 꽤나 혼란스럽군.. 수정이 덜 된 부분들을 빌드 에러를 통해 체크하고 있는데 매우 원시적으로 느껴져서 뭔가 서글프다.

그리고 코드를 서술할 때 "객체"라는 용어가 적절한 용어인가요? 인스턴스라고 불러야하는 건지 아님 그냥 "위임객체" 말고 "대리자" 같은 더 추상적인 이름으로 부르는 게 의사소통에 좋을지 궁금하군요..:confused:

paikwiki committed:
Remove duplicated code for View update

paikwiki commented:
`GameDelegate`의 `gameDidPitch(pitchString:)` 제거 관련해서

`Game` 클래스의 `doPitching(pitchNumber:)`의 동작 때문에 `gameDidPitch(pitchString:)`를 제거하기가 어려워서 우선 둬봤습니다.

```swift
    func doPitching(pitchNumber: Int) {
        guard
            inning.pitching.contains(pitchNumber) == false,
            isOver == false
            else { return }
        inning.pitching.append(pitchNumber)              // A
        gameDidPitch(pitchString: inning.description)    // B
        if inning.isEnded == true {                      // C
            finishCurrentInning()
        }
    }
```

`doPitching(pitchNumber:)`은 `inning.pitching`에 `pitchNumber`를 삽입한 후(A), 바로 뷰를 업데이트 하고(B), 이후에 이닝이 끝났는지 판단하는데(C), A와 C 사이에서 B를 실행하기 위해 델리게이트 메서드를 적용한 거였거든. 얘기를 하다보니 이닝이 끝났는지 판단하는 로직이 `doPitching(pitchNumber:)`안에 있는 게 문제인 것 같기도 하네..

rinndash commented:
내가 코드 리뷰 하는 걸 한참 까먹고 있었네.. 연락이 좀 없다 싶으면 카톡 주십쇼 ㅎㅎ

> Delegate를 적용하려고 하니까 화면에 출력하기 위한 문자열 처리가 말썽이네. inningResultString를 전달해주지 않고 처리하려면 Game의 위임객체 역할을 하는 ViewController가 Answer를 알고 있어야 하잖아. ViewController가 judgePitching(inning: Inning)을 부를 수 없으니 위임객체 역할을 하는 ViewController한테 문자열을 전달해주는 건데, 별로 좋지 않은 방법인데 뾰족한 방법을 잘 모르겠군요...

요거는 좋은 고민인 것 같습니다. 화면에 출력하기 위한 문자열 처리가 어디에서 들어가는 게 좋은가의 문제인데 나는 현재 쓰고 있는 MVC 아키텍처에서는 ViewController에 들어가는 게 맞다고 생각해. `Game`은 이 애플리케이션의 모델이고 `UILabel`이 뷰니까 모델의 상태를 뷰에 적절한 형태로 바꿔주는 역할은 컨트롤러인 `ViewController`가 하는게 맞지. 그러니까 `GameDelegate`에서는 문자열 표현 없이 `Inning`만 넘겨주면 된다고 생각함.

나중에 MVVM으로 아키텍처를 바꾸면 모델과 뷰컨트롤러 사이에 뷰모델이라는 게 생기는데, MVC에서 컨트롤러 역할을 하는 `UIViewController`가 개념적으로는 뷰 레이어에 포함되게 됩니다. 이건 나중에 MVVM 하게 되면 좀 더 자세히 얘기해줄게.

> 그리고 코드를 서술할 때 "객체"라는 용어가 적절한 용어인가요? 인스턴스라고 불러야하는 건지 아님 그냥 "위임객체" 말고 "대리자" 같은 더 추상적인 이름으로 부르는 게 의사소통에 좋을지 궁금하군요..😕

그때 그때 다른데 객체와 인스턴스는 거의 같은 의미로 쓰는 거 같고, 위 상황에서는 위임객체나 대리자보다는 델리게이트라는 영단어를 쓰는게 일반적인듯..

paikwiki commented:
> 내가 코드 리뷰 하는 걸 한참 까먹고 있었네.. 연락이 좀 없다 싶으면 카톡 주십쇼 ㅎㅎ
> 
> > Delegate를 적용하려고 하니까 화면에 출력하기 위한 문자열 처리가 말썽이네. inningResultString를 전달해주지 않고 처리하려면 Game의 위임객체 역할을 하는 ViewController가 Answer를 알고 있어야 하잖아. ViewController가 judgePitching(inning: Inning)을 부를 수 없으니 위임객체 역할을 하는 ViewController한테 문자열을 전달해주는 건데, 별로 좋지 않은 방법인데 뾰족한 방법을 잘 모르겠군요...
> 
> 요거는 좋은 고민인 것 같습니다. 화면에 출력하기 위한 문자열 처리가 어디에서 들어가는 게 좋은가의 문제인데 나는 현재 쓰고 있는 MVC 아키텍처에서는 ViewController에 들어가는 게 맞다고 생각해. `Game`은 이 애플리케이션의 모델이고 `UILabel`이 뷰니까 모델의 상태를 뷰에 적절한 형태로 바꿔주는 역할은 컨트롤러인 `ViewController`가 하는게 맞지. 그러니까 `GameDelegate`에서는 문자열 표현 없이 `Inning`만 넘겨주면 된다고 생각함.
> 
> 나중에 MVVM으로 아키텍처를 바꾸면 모델과 뷰컨트롤러 사이에 뷰모델이라는 게 생기는데, MVC에서 컨트롤러 역할을 하는 `UIViewController`가 개념적으로는 뷰 레이어에 포함되게 됩니다. 이건 나중에 MVVM 하게 되면 좀 더 자세히 얘기해줄게.
> 
> > 그리고 코드를 서술할 때 "객체"라는 용어가 적절한 용어인가요? 인스턴스라고 불러야하는 건지 아님 그냥 "위임객체" 말고 "대리자" 같은 더 추상적인 이름으로 부르는 게 의사소통에 좋을지 궁금하군요..😕
> 
> 그때 그때 다른데 객체와 인스턴스는 거의 같은 의미로 쓰는 거 같고, 위 상황에서는 위임객체나 대리자보다는 델리게이트라는 영단어를 쓰는게 일반적인듯..

코멘트 감사. 이번주에 42서울 과제하느라 바빴습니다ㅠㅠ 다음주에 다른 일로 5일정도 비우게 될 것 같은데 일요일이나 가능하다면 영상통화 한번 합시다.
